Add Cypress tests for UI interactions and form validations

brunonf15 opened the pull request:
### ✅ **Updated Pull Request Description:**  

**Summary:**  
This PR enhances and refines **Cypress tests** to validate key UI functionalities, including form validations, messenger interactions, footer verification, and responsive table behavior.  

**Changes include:**  
- 🛠 **Improved button selection strategies** using CSS attribute selectors instead of relying on `data-testid`.  
- 📩 **Enhanced messenger interactions** (open, close, submit, reset behavior).  
- 🔍 **More robust footer link verification** (ensuring correct attributes and behavior).  
- 📱 **Updated responsive table test** (confirming correct columns are shown/hidden in mobile view).  

**How to Test:**  
1. Run `npm run cypress:open` or `npx cypress run`.  
2. Verify that all tests pass in the Cypress Test Runner.  

**Why?**  
- Improves **test reliability** by using **CSS attribute selectors** instead of relying solely on `data-testid`.  
- Ensures the UI behaves correctly across different scenarios.  
- Helps catch regressions in future updates.  

**Reviewer Notes:**  
- Focus on how I **mapped the buttons without `data-testid`**, using:  
  ```javascript
  cy.get('[class^="Messenger_openCloseButton"]') 
  ```

brunonf15 committed:
Adicionando testes em Cypress ao projeto

wlsf82 commented on .prettierrc:
```diff
@@ -0,0 +1,10 @@
+    "useTabs": false,
+    "bracketSpacing": true,
+    "arrowParens": "always"
+}
```
Recomenda-se deixar uma linha em branco ao final de cada arquivo (independente da extensão).

Ao deixar tal linha em branco, ferramentas como o GitHub não exibem o ícone de `No newline at end of file`.

A propósito, é possível configurar o ESLint para garantir que tal regra seja sempre cumprida.

```suggestion
}

```

wlsf82 commented on cypress.config.js:
```diff
@@ -0,0 +1,7 @@
+  e2e: {
+    baseUrl: 'http://localhost:3000',
+  },
+})
```
Mesma coisa aqui.

wlsf82 commented on cypress/e2e/api/engageAPI.cy.js:
```diff
@@ -0,0 +1,11 @@
+describe('EngageSphere API - Customers', () => {
+  it('Successfully retrieves customers with default filters', () => {
+    cy.request(
+      'http://localhost:3000/customers?page=1&limit=10&size=All&industry=All'
```
Há alguns detalhes a considerar aqui.

Primeiro, visto que a API já possui valores padrão para cada _query string_, passar os valores padrão aqui é desnecessário e torna o código mais complexo do que o necessário.

Segundo, a API roda na porta 3001, não 3000, portanto, o teste está testando a URL errada.

Por fim, evite codificar a URL direto no teste visto que isso impede a execução do mesmo teste contra diferentes ambientes sem a alteração do código do teste propriamente dito.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+describe('EngageSphere - Customer Management UI', () => {
+  beforeEach(() => {
+    cy.visit('/'); // Assuming 'http://localhost:3000/' is set as the baseUrl in cypress.config.js
```
Eu removeria este comentário, até por quê, se uma URL diferente de `http://localhost:3000/` estivesse definida no arquivo de configurações, o `cy.visit` também funcionaria.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.visit('/'); // Assuming 'http://localhost:3000/' is set as the baseUrl in cypress.config.js
+
+    cy.get('[data-testid="size-filter"]').as('sizeFilter');
+    cy.get('[data-testid="accept-button"]').as('acceptButton');
```
A adição deste atributeo `data-testid` é desnecessária visto que tal elemento pode ser identificado da seguinte forma: `cy.contains('button', 'Accep')`.

A propósito, se cada teste aceita os _cookies_ antes de iniciar, o clique em tal botão poderia já estar no _hook_ `beforeEach` em vez de em cada caso de teste.

Ou ainda melhor, os _cookies_ poderia ser definidos programaticamente antes mesmo a visita a página da aplicação, com o comando `cy.setCookie('cookieConsent', 'accepted')`.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+  it('Maintains selected filter after returning from customer details', () => {
+    cy.get('@acceptButton').click();
+
+    // Verify the size filter exists and is visible
```
Evite comentários óbvios.

Comentários devem ser utilizados para explicarem partes do código não óbvias, e ao ler `.should('be.visible')`, já é bastante claro (e óbvio) que é isso que está sendo verificado.

Por fim, é comum comentários ficarem desatualizado por mudanças no código e o esquecimento de atualizar os comentários.

Portanto, só comente código não óbvio.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('@acceptButton').click();
+
+    // Verify the size filter exists and is visible
+    cy.get('@sizeFilter').should('be.visible');
```
Além disso, essa linha é desnecessesária visto que a linha 16 falharia se o elemento não estivesse visível.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    // Verify the size filter exists and is visible
+    cy.get('@sizeFilter').should('be.visible');
+
+    // Select "Small" from the size filter
```
Alias, este é outro comentário óbvio.

E imagino que outros virão, portanto, não vou comentar em todos, pois creio que já deve ter entendido o ponto.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+
+    // Verify that "Small" is now selected
+    cy.get('@sizeFilter').within(() => {
+      cy.get('option:selected').should('have.value', 'Small');
```
Esta verificação é desnecessesária aqui.

O comportamento padrão de qualquer navegador web é que se uma `option` de um `select` for selecionada, seu valor será o selecionado.

O recado aqui é: teste o comportamento da aplicação, não do _browser_.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    });
+
+    // Navigate to the customer details page (clicking on the last cell of the first row)
+    cy.get('tbody > :nth-child(1) > :nth-child(6)').click();
```
Seletores de elementos devem ser legíveis e este não é o caso, portanto, esse não é um seletor aceitável em um script de automação de testes web.

Recomendo o seguinte:

```suggestion
    cy.contains('button', 'View').click();
```

O que acha?

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('tbody > :nth-child(1) > :nth-child(6)').click();
+
+    // Go back to the customer list page
+    cy.get('.Button_buttonContainer__ti4sL > button').click();
```
Seu seletor possui um valor gerado dinamicamente, portanto, ele não é seguro.

Veja uma opção simplificada abaixo para clicar no botão _Back_.

```suggestion
    cy.contains('button', 'Back').click();
```

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('.Button_buttonContainer__ti4sL > button').click();
+
+    // Verify that "Small" is still selected in the size filter
+    cy.get('@sizeFilter').within(() => {
```
Aqui sim você está testando o comportamento da aplicação, ou seja, que o filtro é mantido ao voltar para a listagem de clientes.

No entanto, a verificação podia ser mais simples.

Algo assim:

```js
cy.get('@sizeFilter').should('have.value', 'Small')
```

Uma única linha em vez de três.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    // Verify the footer is present and visible on the page
+    cy.get('[data-testid="footer"]').should('be.visible');
+
+    // Verify the footer contains the correct copyright text
```
Veja abaixo um conjunto de verificações simplificadas para o _footer_ da aplicação.

Compare com a sua implementação e responda para si mesmo, "qual delas é mais simples?"

Lembrando que um bom design é um design que simplifica o código.

```js
cy.contains('p', 'Copyright 2025 - Talking About Testing')
  .should('be.visible')
cy.contains('a', 'Podcast')
  .should('be.visible')
  .and('have.attr', 'href', 'https://open.spotify.com/show/5HFlqWkk6qtgJquUixyuKo')
  .and('have.attr', 'target', '_blank')
  .and('have.attr', 'rel', 'noopener noreferrer')
cy.contains('a', 'Courses')
  .should('be.visible')
  .and('have.attr', 'href', 'https://talking-about-testing.vercel.app/')
  .and('have.attr', 'target', '_blank')
  .and('have.attr', 'rel', 'noopener noreferrer')
cy.contains('a', 'Blog')
  .should('be.visible')
  .and('have.attr', 'href', 'https://talkingabouttesting.com')
  .and('have.attr', 'target', '_blank')
  .and('have.attr', 'rel', 'noopener noreferrer')
cy.contains('a', 'YouTube')
  .should('be.visible')
  .and('have.attr', 'href', 'https://youtube.com/@talkingabouttesting')
  .and('have.attr', 'target', '_blank')
  .and('have.attr', 'rel', 'noopener noreferrer')
```

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('@acceptButton').click();
+
+    // Verify that the element is empty
+    cy.get('[data-testid="name"]').should('be.empty');
```
Esta verificação sempre irá passar, mesmo que algo esteja digitado no campo.

A _assertion_ `be.empty` dever ser usada para estrutura de Array.

Por exemplo: `expect([]).to.be.empty`.

Para verificar que um input está vazio, o correto é verificar que sua propriedade `value` seja igual a uma _string_ vazia.

Ou seja, cy.get('[data-testid="name"]').should('have.value', '');

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('[data-testid="name"]').should('be.empty');
+
+    // Verify that the element contains "Hi there"
+    cy.get('[data-testid="greeting-header"]').should('contain', 'Hi there');
```
Uma verificação simplificada, sem a necessidade de `data-testid` seria a seguinte:

```js
cy.contains('h2', 'Hi there').should('be.visible')
```

A propósito, perceba a verificação pela visibilidade do elemento.

Um elemento pode existir no DOM e não estar visível, portanto, é importante verificar por sua visibilidade.

wlsf82 replied:
Além disso, perceba como o código fica mais curto.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('[data-testid="name"]').type('Joe');
+
+    // Verify that the element contains "Hi there"
+    cy.get('[data-testid="greeting-header"]').should('contain', 'Hi Joe');
```
A mesma coisa se aplica aqui.

Poderia ser:

```suggestion
    cy.contains('h2', 'Hi Joe').should('be.visible')
```

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+
+  it('Exibe o cabeçalho com um título, alternador de tema e um campo de entrada de texto', () => {
+    // Verify the page title is "EngageSphere"
+    cy.title().should('eq', 'EngageSphere');
```
Aqui a verificação devia ser o `h1`, não do `title`.

Ou seja, `cy.contains('h1', 'EngageSphere').should('be.visible')`.

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    // Verify that the messenger is open (element is visible)
+    cy.get('#messenger-name').should('be.visible');
+
+    // Close the messenger
```
Você percebe como todos estes comentários são desnecessários e óbvios?

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+    cy.get('#messenger-name').should('be.focused');
+
+    // Try to submit the form without filling any fields
+    cy.get('[data-testid="send-button"]').click();
```
Este clique não deveria ser necessário para verificar a obrigatoriedade dos campos.

Em vez disso, você poderia fazer algo assim:

```js
cy.get('#messenger-name').should('have.attr', 'required')
```

wlsf82 commented on cypress/e2e/e2e/engageE2e.cy.js:
```diff
@@ -0,0 +1,228 @@
+      .and('contain', 'Your message has been sent.');
+
+    // Wait 3000 for the message to disappear (assuming it auto-hides)
+    cy.wait(3000); // Adjust this time if needed based on your app's behavior
```
Em vez de esperar explicitamente por 3 segundos, você poderia usar as funcionalidades `cy.clock()` para congelar o relógio do navegador antes de visitar a página e então a funcionalidade `cy.tick(3000)` para avançar o relógio em 3 segundos, sem perder os três segundos, tornando o teste mais rápido.

Lembre-se. Testes automatizados devem fornecer feedback rápidos aos times, e _waits_ explícitos tornam os testes lentos.

wlsf82 commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,25 @@
+// ***********************************************
```
Não recomenda-se deixar as pastas `cypress/fixtures/` e `cypress/support/` no projeto quando estas não estão em uso.

## Por que tal prática fere um bom design de testes?

Tal prática aumenta a complexidade do projeto visto que é mais código para ler e entender.

## Sugestão com um melhor design

Caso os arquivos das pastas citadas acima não estejam em uso, é recomendado deletar
tais pastas por completo e configurar o Cypress para ignorá-los, conforme demonstrado abaixo.

```js
const { defineConfig } = require('cypress')

module.exports = defineConfig({
  e2e: {
    baseUrl: '...',
    supportFile: false,
    fixturesFolder: false,
  },
})

```

Dessa forma, ao inicializar o Cypress, o mesmo não irá re-gerar tais arquivos e pastas, a base
de código estará menor e mais simples, que é exatamente o que buscamos com um bom
design (simplicidade).

wlsf82 commented on frontend/src/components/Button/index.js:
```diff
@@ -4,11 +4,16 @@ const Button = ({
   text = '',
-  onClick = () => {}
+  onClick = () => {},
+  dataTestId = ''
```
Recomendo reverter as mudanças no frontend onde atributos `data-testid` foram adiconados, visto que são desnecessários em tais botões, onde os mesmos podem ser facilmente encontrados com o comando `cy.contains('button', 'Texto do botão')`

wlsf82 replied:
Pense duas vezes antes de alterar o código do frontend, afinal, estamos aqui para testar que a aplicação funciona, e se alteramos o código da aplicacão, o que garante que não quebramos nada?

wlsf82 commented on frontend/src/components/Greeting/index.js:
```diff
@@ -5,7 +5,9 @@ const Greeting = ({ name }) => {
   return (
     <>
-      <h2>Hi {name || 'there'}! {name === 'Squirrel' ? <Squirrel /> : null} It is {today.toLocaleDateString("en-US", options)}</h2>
+      <h2 data-testid="greeting-header">
```
O mesmo vale aqui.

O _greeting_ pode ser facilmente encontrado sem um atributo `data-testid`.

Assim:

```js
cy.contains('h2', 'Hi ...').should('be.visible')
```

wlsf82 commented on frontend/src/components/Messenger/index.js:
```diff
@@ -1,37 +1,37 @@
-import { useState, useRef, useEffect } from 'react'
-import { MessageCircle, X, Send } from 'lucide-react'
+import { useState, useRef, useEffect } from 'react';
```
E essas mudanças?

Creio que sejam devido ao uso do prettier, mas se um certo padrão estava sendo usado (o qual era sem ponto-e-vírgula no final das expressões), o que te fez mudar tal padrão?

wlsf82 replied:
Recomendo reverter todas as mudanças no código do frontend.

A idéia aqui é testar que o frontend funciona como está, sem alterá-lo.

wlsf82 commented on package.json:
```diff
@@ -17,7 +17,8 @@
   "devDependencies": {
-    "@babel/plugin-proposal-private-property-in-object": "^7.21.11"
+    "@babel/plugin-proposal-private-property-in-object": "^7.21.11",
+    "cypress": "^14.1.0"
```
Que tal fazer o upgrade do Cypress para a versão 14.2.0, a qual foi lançada no dia 12/03/2025?

**Fonte:** https://docs.cypress.io/app/references/changelog#14-2-0

wlsf82 reviewed (changes requested):
Bom trabalho @brunonf15! 🎉 

No entanto, deixei uma série de comentários ao longo do seu código, os quais creio que lhe ajudarão a melhorar o design do mesmo.

> **Obs.:** Só faça novas alterações depois do encontro de hoje da Test Design Masterclass.

brunonf15 committed:
Updating the README.md

wlsf82 commented on .prettierrc:
```diff
@@ -0,0 +1,9 @@
+{
```
Precisa mesmo do prettier?

Que tal em vez dele configurar o ESLint?

brunonf15 replied:
Não precisa e eu optei por remover o arquivo porque não ia ter tempo de configurar o ESLint

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+
+### Installing Cypress
+
+1. Ensure you have `Node.js` and `npm` installed as per the pre-requirements above.
```
Esta informação não é necessária visto que o `README.md` já possui uma sessão de pré-requisitos que comenta sobre a necessidade do Node.js e npm.

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+
+1. Ensure you have `Node.js` and `npm` installed as per the pre-requirements above.
+2. Navigate to the root directory of your project.
+3. Install Cypress as a development dependency by running:
```
Visto que o `cypress` é listado na sessão de `devDependencies` do arquivo `package.json`, para instalá-lo, basta rodar `npm install`, não `npm install cypress --save-dev`.

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+3. Install Cypress as a development dependency by running:
+
+```bash
+npm install cypress --save-dev
````
```suggestion
npm install
```

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+
+The test files are organized as follows:
+
+- **API Tests**: Located in `cypress/e2e/api/engageAPI.cy.js`.
```
Visto que o arquivo está na pasta `api/`, o arquivo não precisa ter esse pós-fixo.

Podia ser `cypress/e2e/api/engageSphere.cy.js`

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+The test files are organized as follows:
+
+- **API Tests**: Located in `cypress/e2e/api/engageAPI.cy.js`.
+- **GUI Tests**: Located in `cypress/e2e/e2e/engageGUI.cy.js`.
```
E aqui, devia ser `cypress/e2e/gui/engageSphere.cy.js`

wlsf82 replied:
`e2e/e2e/` fica confuso.

wlsf82 replied:
Eu já havia comentado isso no último encontro.

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+- **API Tests**: Located in `cypress/e2e/api/engageAPI.cy.js`.
+- **GUI Tests**: Located in `cypress/e2e/e2e/engageGUI.cy.js`.
+
+
```
Uma linha em branco basta.

```suggestion
```

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+- **Interactive Mode**: Launch Cypress Test Runner with:
+
+```bash
+npx cypress open
````
Em vez de usar `npx`, recomendo definir scripts no arquivo `package.json` e aqui referenciar os mesmos.

Exemplo:

```json
"scripts": {
  "cy:open": "cypress open",
  "test": "cypress run"
}

```

E então, na documentação podia haver algo como:

To run the tests in interactive mode, run `npm run cy:open`.

And to run the tests in headless mode, run `npm test`.

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+- **Headless Mode**: Run only this specific test file:
+
+```bash
+npx cypress run --spec "cypress/e2e/api/engageAPI.cy.js"
````
Você também poderia ter scripts no `package.json` que fazem isso.

Algo assim:

```json
"scripts": {
  "cy:open": "cypress open",
  "test": "cypress run",
  "test:api": "cypress run --spec 'cypress/e2e/api/engageSphere.cy.js'",
  "test:gui": "cypress run --spec 'cypress/e2e/gui/engageSphere.cy.js'"
}

```

E na documentação, podia ser assim:

To run the API tests in headless mode, run `npm run test:api`.

To run the GUI tests in headless mode, run `npm run test:gui`.

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
 Made with ❤️ by [Walmyr](https://walmyr.dev).
+Tests by Bruno Figueiredo
+
+<div style="text-align: center">⁂</div>
```
O que isso faz?

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,116 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+
+<div style="text-align: center">⁂</div>
+
+[^1]: https://pplx-res.cloudinary.com/image/upload/v1742568908/user_uploads/VrYjtGtFTuPLZXf/image.jpg
```
E isso?

wlsf82 commented on cypress/e2e/api/engageAPI.cy.js:
```diff
@@ -0,0 +1,11 @@
+      `${Cypress.env("apiUrl")}/customers`
+    ).as('api');
+
+    cy.get('@api').should((response) => {
```
Recomendo usar desestruturação de objetos aqui.

Assim:

```js
cy.get('@getCustomers').should(({ status }) => {
  expect(status).to.equal(200);
});
```

O que acha?

wlsf82 commented on cypress/e2e/api/engageAPI.cy.js:
```diff
@@ -0,0 +1,11 @@
+  it('Successfully retrieves customers with default filters', () => {
+    cy.request(
+      `${Cypress.env("apiUrl")}/customers`
+    ).as('api');
```
Que tal um alias mais descritivo, como `getCustomers`?

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+describe('EngageSphere - Customer Management UI', () => {
+  beforeEach(() => {
+    cy.visit('/');
+    //cy.setCookie('cookieConsent', 'accepted')
```
Mova o `cy.setCookie` para antes do `cy.visit` que vai funcionar.

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+describe('EngageSphere - Customer Management UI', () => {
```
Customer Management?

Não seria Customers List?

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+    cy.contains('button', 'Accep').click();
+  });
+
+  it('Maintains selected filter after returning from customer details', () => {
```
Veja abaixo uma versão que separa as diferentes partes do testes por uma linha em branco, para separar pré-condição, ações e resultado esperado.

```js
it('Maintains selected filter after returning from customer details', () => {
  // Arrange
  cy.get('[data-testid="size-filter"]')
    .as('sizeFilter');
    .select('Small');
  cy.contains('button', 'View').click();

  // Act
  cy.contains('button', 'Back').click();

  // Assert
  cy.get('@sizeFilter').should('have.value', 'Small');
});

```

wlsf82 replied:
Mesmo que não houvessem os comentários do padrão AAA, seria possível facilmente identificar as diferentes partes do teste, visto que há uma linha em branco entre cada parte.

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+    cy.get('@sizeFilter').should('have.value', 'Small');
+  });
+
+  it('Verify footer links', () => {
```
👍🏻

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+      .and('have.attr', 'rel', 'noopener noreferrer');
+  });
+
+  it('Empty Hi there', () => {
```
Recomendo usar a descrição da sugerida na lista de casos de teste.

O mesmo vale para o próximo teste.

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+  });
+
+  it('Hi Joe there', () => {
+    cy.get('[data-testid="name"]').type('Joe');
```
Aqui vai uma pequena melhoria, que separa ação e resultado esperado.


```js
it('Hi Joe there', () => {
  cy.get('[data-testid="name"]').type('Joe');

  cy.contains('h2', 'Hi Joe').should('be.visible');
});
```

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+
+  it('Exibe o cabeçalho com um título, alternador de tema e um campo de entrada de texto', () => {
+    cy.contains('h1', 'EngageSphere').should('be.visible');
+    cy.get('[class^="ThemeToggle_button"]').should('be.visible');
```
Que tal criar um comando customizado conforme abaixo?

```js
Cypress.Commands.add('getByClassThatStartsWith', classPart => {
  cy.get(`[class^="${classPart}"]`)
})
```

E então, aqui (e em outros lugares), você poderia chamar `cy.getByClassThatStartsWith('ThemeToggle_button')`.

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+    cy.get('[data-testid="name"]').should('be.visible');
+  });
+
+  it('Verificar que o preenchimento do campo nome é obrigatório', () => {
```
Esse teste é desnecessário visto que o próximo já verifica esse campo, além dos outros.

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+    cy.get('#message').should('have.attr', 'required');
+  });
+
+  it('Mostra e oculta uma mensagem de sucesso ao enviar o formulário do messenger', () => {
```
Você percebe como fica difícil identificar as pré-condições, ações e resultados esperados quando está tudo colado um no outro, sem uma linha em branco para separar tais partes do teste?

wlsf82 commented on cypress/e2e/e2e/engageGUI.cy.js:
```diff
@@ -0,0 +1,113 @@
+    cy.get('[class^="Messenger_success"]').should('not.exist');
+  });
+
+  it('Limpa todos os campos do formulário ao fechar e reabrir o messenger', () => {
```
O mesmo aqui.

wlsf82 reviewed (changes requested):
Bom trabalho @brunonf15! 👏🏻 👏🏻 

No entanto, deixei novos comentários, os quais creio que lhe ajudarão a melhorar ainda mais o design dos testes.

> Lembre-se de aplicar as mudanças sugeridas somente após o terceiro encontro da _Test Design Masterclass_ da TAT.

brunonf15 committed:
Fixing setCookie

brunonf15 committed:
Fixing PR suggestions

Adding new scripts to run API and gui tests
Adding new scripts to run Cypress tests
Rename api and gui tests
Updating the readme

brunonf15 committed:
Adding better name to the alias

brunonf15 committed:
Fixing the PR suggestions

Renaming the tests cases acording to the list
Adding some extra spactes to improve the readbility of the tests

brunonf15 committed:
Fixing the latest suggestions

Using a command
Adding more spaces to improve the readability of the code

brunonf15 committed:
Adding more API tests

brunonf15 committed:
Adding a new test to evidence the bug

brunonf15 committed:
Removing prettier

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,104 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+### Installing Cypress
+
+1. Install Cypress as a development dependency by running:
+```bash
````
```suggestion

```bash
```

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,104 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+```bash
+npm install 
+```
+
````
Uma linha em branco já é o suficiente.

```suggestion
```

wlsf82 commented on README.md:
```diff
@@ -24,6 +24,104 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+
+- **API Tests**: Located in `cypress/e2e/api/engageSphere.cy.js`.
+- **GUI Tests**: Located in `cypress/e2e/gui/engageSphere.cy.js`.
+
```
O mesmo aqui.

```suggestion
```

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,104 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+```bash
+npm test
+```
+
````
```suggestion
```

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,104 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+```bash
+npm run test:api
+```
+
````
```suggestion
```

wlsf82 commented on README.md:
````diff
@@ -24,6 +24,104 @@ Read the following [doc](./docs/TestEnvironment.md) to install and start the bac
+```bash
+npm run test:gui
+```
+
````
```suggestion
```

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+    cy.request(`${Cypress.env("apiUrl")}/customers`).as('getCustomers');
+
+    cy.get('@getCustomers').should((response) => {
+      expect(response).to.have.property('status', 200);
```
Outra alternativa (usando desestruturação de objetos) seria:

```js
cy.get('@getCustomers').should(({ status }) => {
  expect(status).to.eq(200);
});

```

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: {
+        page: 2,
+        limit: 1,
```
Se o teste é para a _query string_ `page`, não há necessidades de passar `limit`, `size` e `industry`, ainda mais passando-os com seus valores padrão.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+        size: 'All',
+        industry: 'All'
+      }
+    }).then((response) => {
```
Usando desestruturação de objetos, ficaria assim:

```js
}).then(({ body, status }) => {
  expect(status).to.eq(200);
  expect(body).to.have.property('pageInfo');
  expect(body.pageInfo).to.deep.include({
  currentPage: 2,
});

```

Perceba que dessa forma, não se repete `response.`

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      method: 'GET',
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: {
+        page: 2,
```
Se o teste é para a _query string_ `limit`, não há necessidade de passar `page`, `size` e `industry`.

Foque no que está sendo testado e deixe a API usar seus padrões para o resto.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      }
+    }).then((response) => {
+      const totalCustomers = response.body.pageInfo.totalCustomers;
+      const expectedTotalPages = Math.ceil(totalCustomers / limit);
```
Não acho uma boa haver cálculos no meio de um teste.

Veja minha implementação abaixo:

```js
it('filters by limit of customers', () => {
  const resultsPerPageLimit = [5, 10, 20, 50]
  const totalPagesPerLimit = [10, 5, 3, 1] // Supposing there are 50 clients in the database.

  resultsPerPageLimit.forEach((limit, index) => {
    cy.request('GET', `${CUSTOMERS_API_URL}?limit=${limit}`).as('getLimittedCustomers')

    cy.get('@getLimittedCustomers')
      .its('body.customers')
      .should('have.length', limit)
    cy.get('@getLimittedCustomers')
      .its('body.pageInfo.totalPages')
      .should('be.eq', totalPagesPerLimit[index])
  })
})

```

wlsf82 replied:
Explicarei com detalhes esse teste no encontro de amanhã.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+  it('Trata requisições inválidas (por exemplo, página negativa)', () => {
+    cy.request({
+      method: 'GET',
+      url: `${Cypress.env("apiUrl")}/customers`,
```
Visto que esta URL se repete em vários testes, que tal criar uma variável chamada `CUSTOMERS_API_URL` no início do arquivo, a qual poderia ser reaproveitada ao longo de todos os testes sem repetir `${Cypress.env("apiUrl")}/customers`?

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+    cy.request({
+      method: 'GET',
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: { page: -1, limit: 5, size: 'All', industry: 'All' },
```
Basta passar `page`. Deixe a API usar seus _defaults_ para o resto.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: { page: -1, limit: 5, size: 'All', industry: 'All' },
+      failOnStatusCode: false
+    }).then((response) => {
```
Use desestruturação de objetos, conforme exemplificado anteriormente, dessa forma, evita-se repetir `response.`.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+    });
+  });
+
+  it('Trata requisições inválidas (por exemplo, limite negativo)', () => {
```
Basta passar `limit`. Deixe a API usar seus _defaults_ para o resto.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: { page: 1, limit: -5, size: 'All', industry: 'All' },
+      failOnStatusCode: false
+    }).then((response) => {
```
Desestruturação de objetos.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+    cy.request({
+      method: 'GET',
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: { page: 'abc', limit: 5, size: 'All', industry: 'All' },
```
O mesmo aqui. Passe só a _query string_ alvo do teste.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      url: `${Cypress.env("apiUrl")}/customers`,
+      qs: { page: 'abc', limit: 5, size: 'All', industry: 'All' },
+      failOnStatusCode: false
+    }).then((response) => {
```
Use desestruturação de objetos.

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      qs: {
+        page: 1,
+        limit: 5,
+        size: 'InvalidSize', // Valor não suportado
```
Precisa mesmo desse comentário? Seu valor já deixa isso claro, não?

```suggestion
        size: 'InvalidSize',
```

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+        page: 1,
+        limit: 5,
+        size: 'All',
+        industry: 'InvalidIndustry' // Valor não suportado
```
```suggestion
        industry: 'InvalidIndustry'
```

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+      expect(response.body).to.have.property('error', 'Unsupported industry value. Supported values are All, Logistics, Retail, Technology, HR, and Finance.');
+    });
+  });
+
```
```suggestion
```

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+    });
+  });
+
+
```
```suggestion
```

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,141 @@
+import '../support/commands'
```
Não existe a necessidade desse import.

Qualquer comando customizado automaticamente fica disponível no objeto global `cy.`

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,141 @@
+    cy.get('[data-testid="name"]').should('be.visible');
+  });
+
+  it('Abre e fecha o messenger', () => {
```
Aqui podia ser:

```js
it('Abre e fecha o messenger', () => {
  // Act
  cy.getByClassThatStartsWith('Messenger_openCloseButton').click();

  // Assert
  cy.get('#messenger-name').should('be.visible');

  // Act
  cy.getByClassThatStartsWith('Messenger_openCloseButton').click();

  // Assert
  cy.get('#messenger-name').should('not.exist');
});

```

O que acha? Pra separar melhor as diferentes partes do teste?

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,141 @@
+    cy.getByClassThatStartsWith('Messenger_success')
+      .should('be.visible')
+      .and('contain', 'Your message has been sent.');
+    cy.tick(3000);
```
```suggestion
    cy.tick(3000);

```

Deixando uma linha pra separar o _assert_.

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,141 @@
+  });
+
+  it('Limpa todos os campos do formulário do messenger ao preenchê-los, fechar o messenger e abri-lo novamente', () => {
+    cy.getByClassThatStartsWith('Messenger_openCloseButton').click();
```
```suggestion
    // Arrange
    cy.getByClassThatStartsWith('Messenger_openCloseButton').click();
    
    // Act
```

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,141 @@
+    cy.getByClassThatStartsWith('Messenger_openCloseButton').click();
+    cy.getByClassThatStartsWith('Messenger_openCloseButton').click();
+
+    cy.get('#messenger-name').should('have.value', '');
```
E então:

```suggestion
    // Assert
    cy.get('#messenger-name').should('have.value', '');
```

wlsf82 commented on cypress/e2e/gui/engageSphere.cy.js:
```diff
@@ -0,0 +1,141 @@
+    cy.contains('th', 'Size').should('not.be.visible');
+  });
+  
+  it('Bug - Edit name bloquado apos filtro não retornar dados', () => {
```
A descrição do caso de teste deve ser do comportamento sendo testado, não a descrição do bug.

Algo como:

```suggestion
  it('Reativa o campo de entrada ao retornar de um filtro de estado vazio para um não vazio', () => {
```

wlsf82 commented on cypress/e2e/support/commands.js:
```diff
@@ -0,0 +1,3 @@
+Cypress.Commands.add('getByClassThatStartsWith', classPart => {
+    cy.get(`[class^="${classPart}"]`)
+  })
```
```suggestion
})

```

wlsf82 reviewed (changes requested):
Bom demais @brunonf15! Com ajustes de pequenos detalhes essa suíte de testes ficará 🔝 

👏🏻 👏🏻 👏🏻

wlsf82 commented on cypress/e2e/api/engageSphere.cy.js:
```diff
@@ -0,0 +1,135 @@
+    });
+  });
+
+  it('Trata requisições inválidas (por exemplo, página negativa)', () => {
```
Que tal criar também casos de testes para os cenários onde `page` e `limit` são iguais a `0`?

wlsf82 commented on cypress.config.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = defineConfig({
+  e2e: {
+    baseUrl: "http://localhost:3000",
+    supportFile: false,
```
Visto que você criou o comando customizado, remova esta configuração, crie o arquivo `cypress/support/e2e.js` e dentro dele importe o arquivo `cypress/support/commands.js`.
